### DISCUSSION TITLE
Recent experiments with ddc other settings

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#234] [Archived]
<USER>>>> nmstoker<br/> <DATE> [July 14, 2020, 4:29pm]<br/> <br/> I wanted to post some details of recent runs, to see how they were<br/> consistent or different for others doing similar runs.<br/> <br/> Prior to the last few weeks, I'd been working on a copy of the TTS repo<br/> from back in late February, and with that I'd got to the point where I<br/> had pretty good results (in general and in particular when used with the<br/> external MelGAN from<br/> [ParallelWaveGAN](https://github.com/erogol/ParallelWaveGAN).<br/> <br/> However I did have some issues with longer sentences and I was keen to<br/> try out the most recent developments with DDC and the integrated<br/> vocoder. Therefore I updated with the latest code from dev and tried<br/> various runs.<br/> <br/> Initially I'll post results here relating to the main TTS model and I<br/> plan to add more detail covering the integrated vocoder shortly (as that<br/> needs more runs first, so it may be about a week or so).<br/> <br/> Obviously with using the dev code, it's a little more 'at the forefront'<br/> and I knew to expect a few bumps on the way!<br/> <br/> # Tensorboard charts<br/> <br/> **Note:** I've turned off the last two as they're for the integrated<br/> vocoder and the differing chart types will confuse things here)<br/> <br/> # Summary of settings<br/> <br/> Colour on charts Not on Chart Orange Mid Blue Dark Red Light Blue Pink Green Grey Orange (blip at 300k)<br/> ------------------ -------------- ---------------- ---------------- ---------------- ---------------- ---------------- ---------------- --------------------- -----------------------<br/> **Late Feb** **July *slash* -- *slash* >** <br/> **Phon set** Full Redux Full Full Redux Redux Redux Full (ignore name!) Full (ignore name!)<br/> **DDC** No Yes, ddc_r = 1 Yes, ddc_r = 1 Yes, ddc_r = 1 Yes, ddc_r = 7 Yes, ddc_r = 7 Yes, ddc_r = 7 No No<br/> **fft_size** 1025 1024 1024 1024 1024 1024 1024 1024 1024<br/> **win_len** N/A 1024 1024 1024 1024 1024 1024 1024 1024<br/> **hop_len** N/A 256 256 256 256 256 256 256 256<br/> **frm_len** 50 null null null null null 50 null null<br/> **frm_shft** 12.5 null null null null null 12.5 null null<br/> **Stats path** N/A Yes Yes Yes Yes No No Yes Yes<br/> **Pre-emph** 0.98 0.00 0.00 0.00 0.00 0.00 0.98 0.00 0.00<br/> **Ref Lvl Db** 25 0 0 0 20 25 25 25 25<br/> **Mel FMin** 0.0 50.0 50.0 50.0 50.0 40.0 40.0 40.0 40.0<br/> **Mel FMax** 8k 7.9k 7.9k 7.9k 8k 8k 8k 8k 8k<br/> **Spec Gain** N/A 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0<br/> **Steps** 580k 170k 60k 40k 70k *slash* >10k *slash* >40k 370k *slash* ~0<br/> <br/> # Comments<br/> <br/> ## Before first successful runs<br/> <br/> Initially had difficulties getting it to run without running out of<br/> memory almost immediately on my GTX 1080Ti.<br/> <br/> At that point I hadn't realised that I'd left the ddc_r = 1 value that<br/> was in the config I'd taken from the CoLab (*in hindsight: clearly best<br/> to start with config from the repo instead*<br/> I just assumed that using two decoders for DDC would need much more<br/> memory (likely true but not at this level).<br/> <br/> I tried reducing the batch sizes used for the regular decoder. That only<br/> worked when they were all down to 16 (which has other negative<br/> implications but I decided to press on at that point).<br/> <br/> ## 'Orange 1'<br/> <br/> My first July run. The audio was *really* distorted on the peaks, with<br/> that aspect not improving with time as I ran it longer.<br/> <br/> Log output was indicating a lot of warnings of '**audio amplitude out of<br/> range, auto clipped**' - those have come up for me before but saw a lot<br/> more than usual and they didn't improve given time.<br/> <br/> I had applied the stats_path normalisation. Owing to the distortion I<br/> wondered if it was causing issues with my dataset, so I tried varying a<br/> number of things with subsequent runs.<br/> <br/> Was also interested in the idea that not all phonemes are used in<br/> English, so I experimented with cutting the phoneme character list<br/> defined in config.json to be just those strictly needed for the model.<br/> In the next run I returned to the full phoneme set to see if this had<br/> somehow caused or exacerbated the distortion.<br/> <br/> ## Mid Blue<br/> <br/> Although I returned to using the full set of phonemes, I had<br/> accidentally left in place the cached phonemes from the reduced set, so<br/> this was a useless run - the audio was jibberish (which I only realised<br/> when I checked hours later)<br/> <br/> ## Dark Red<br/> <br/> Audio still fairly distorted; stopped it early as realised the ddc_r<br/> issue (fixed in next run)<br/> <br/> ## Light Blue<br/> <br/> Fixed the ddc_r issue, setting it to 1.<br/> <br/> Alignment 2 (ie for ddc) was markedly better now as expected.<br/> <br/> Audio still badly distorted<br/> <br/> ## Pink and Green<br/> <br/> These both had problems that I couldn't pin down. The audio in both was<br/> almost non-existent, aside from background noise. The charts for Pink<br/> and Green looked odd too.<br/> <br/> Pink was worse as it had stopping failures (so samples ran on for 59<br/> seconds), with Green they returned to more normal time ranges (eg 2-3<br/> seconds). They both had the reduced phoneme set.<br/> <br/> With Green I tried the frame length and shifts I'd had in my previous<br/> best run (from late Feb-May series of runs).<br/> <br/> ## Grey and Orange 2<br/> <br/> For Grey I gave up on the phoneme idea.<br/> <br/> I let it run for quite some time and Grey is the best of my July runs so<br/> far.<br/> <br/> It's still clearly distorted, but if you ignore the peaks, it has quite<br/> good qualities (although a long way from usable)<br/> <br/> I was going to continue training with Orange 2 but decided I had better<br/> explore whether the integrated vocoder could somehow counteract the<br/> distortions, so stoped Orange 2 and switch to look at the vocoder. I<br/> plan to write that up once I've done more runs and having findings to<br/> share.<br/> <br/> I'm still interested to see if I can figure out what to do around<br/> normalisation of the audio. I wonder if for my dataset somehow the<br/> stats_path approach is too much. I hope to try plotting some of my audio<br/> samples *before* and *after* to explore it further.<br/> <br/> I know voice audio can be aymmetric so I wonder if the higher amplitude<br/> side is being amplified enough to go out of range whilst the lower side<br/> is still well within range (this is part of the motivation behind my<br/> desire to plot the audio samples).<br/> <br/> Hope that's of interest. If you've had similar (or different!)<br/> experiences, would be great to hear about them.<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/recent-experiments-with-ddc-other-settings]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [July 15, 2020, 12:58pm]<br/> <br/> By the way: if viewing this on a large screen desktop and you find the<br/> table's a bit squashed (as the desktop mode of the forum's layout is<br/> rather narrow) you can view it in responsive mode, which should use the<br/> full width, with this link where the parameter **?mobile_view=1** is<br/> appended:<br/> <br/> ![](../../user_avatar/discourse.mozilla.org/nmstoker/40/30582_2.png){.avatar<br/> settings](https://discourse.mozilla.org/t/recent-experiments-with-ddc-other-settings/63787?mobile_view=1)<br/> drop-close='true'<br/> .box}<br/> <br/> > I wanted to post some details of recent runs, to see how they were<br/> > consistent or different for others doing similar runs. Prior to the<br/> > last few weeks, I'd been working on a copy of the TTS repo from back<br/> > in late February, and with that I'd got to the point where I had<br/> > pretty good results (in general and in particular when used with the<br/> > external MelGAN from<br/> > [ParallelWaveGAN](https://github.com/erogol/ParallelWaveGAN). However<br/> > I did have some issues with longer sentences and I was keen to try out<br/> > the most recent developments with DDC and t...<br/> <br/> **NB:** You'll need to resubmit the page address in the browser address<br/> bar (ie it doesn't work just by clicking the link)<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [August 2, 2020, 4:00pm]<br/> <br/> **Quick update:** After a fair bit of experimenting, I've got some<br/> pretty good results with the integrated Vocoder with an example reading<br/> of a recipe here: <https://youtu.be/6gECEaZhtio><br/> <br/> The TTS model used here wasn't trained with DDC but it manages to be<br/> fairly stable all the same. The Vocoder ran for 1,000,000 steps and is<br/> pretty smooth - I plan to do a little more experimentation to see if I<br/> can fine-tune just a bit more quality out of it.<br/> <br/> Overall, I'm really pleased. There is still the occasional glitch with a<br/> word. Generally glitches seem most prevalent with words when they are<br/> used in speech patterns that didn't come up in my training. When correct<br/> phonemes are supplied by espeak-ng it can easily go beyond the original<br/> vocabulary fairly well, it's more the way the words are used that seems<br/> the issue. A limit of my original TTS model training was on the length<br/> of short words used and that definitely shows - it can't say 'Good.' on<br/> its own, but is fine with it in a sentence!<br/> <br/> Lists, questions, imperatives and splitting up clauses seem to be<br/> reasonably well handled too.<br/> <br/> [Archived Post]<br/>
